src/libostree/ostree-repo.c

```diff
@@ -660,7 +718,11 @@ ostree_repo_auto_lock_push (OstreeRepo          *self,
 {
   if (!ostree_repo_lock_push (self, lock_type, cancellable, error))
     return NULL;
-  return (OstreeRepoAutoLock *)self;
+
+  OstreeRepoAutoLock *auto_lock = g_new0 (OstreeRepoAutoLock, 1);
```

Comment:
Could the need to heap-allocate be avoided by making the `OstreeRepoAutoLock` struct public (but opaque), and having callers stack-allocate it?

Or, if that’s too intrusive, you could use `g_alloca()` here and require that `OstreeRepoAutoLock` is always used with `g_autoptr()` in a single scope.

Reply:
I did want to stack allocate but couldn't figure out a way that would be as nice to use as `g_autoptr()`. Currently you can do:
```
g_autoptr(OstreeRepoAutoLock) lock = ostree_repo_auto_lock_push(repo, ...);
if (!lock)
  return FALSE;
```
With the stack allocation, I think it would be:
```
g_auto(OstreeRepoAutoLock) lock = { 0, };
if (!ostree_repo_auto_lock_push(repo, &lock, ...))
  return FALSE;
```
I guess that's not so bad. You could probably provide an initializer macro, too. I could go either way there.

Although I can't think of a reason why you'd want to use it outside of a single scope, requiring it seems a bit intrusive.

Reply:
> I guess that's not so bad. You could probably provide an initializer macro, too. I could go either way there.

Yeah, or you could change `ostree_repo_auto_lock_push()` to return an `OstreeRepoAutoLock` structure by value. Should be fine, since it’s only two pointers.

Reply:
That's how I was thinking about it initially, but then how do you detect errors? Particularly if the structure is opaque. Unlike `autoptr` where you can use `NULL`, here you'd have to check if one of the members was in an error state since you'd actually have to return a structure. Even if it's not opaque it would be kind of ugly to use it that way.

Reply:
You could check the `error` out argument. But true, it is quite idiosyncratic. :shrug:

Reply:
I implemented a stack allocated version in https://github.com/dbnicholson/ostree/commit/bd0fb98597ec29a839f67cb75416107368b9cb48. I'm pull that in here if that's the preferred API.